### PR TITLE
miniupnpd: add service

### PR DESCRIPTION
**Pull request by jgillich**
Not sure if there are any more common config options to add.

### CODE REVIEW

File: nixos/modules/services/networking/miniupnpd.nix [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      listening_ip=${range}
+    '') cfg.internalIPs}
+
+    ${cfg.appendConfig}
+  '';
+in
+{
+  options = {
+    services.miniupnpd = {
+      enable = mkOption {
```
Comment (jagajaga) [inline nixos/modules/services/networking/miniupnpd.nix]: Please use `mkEnableOption`.

File: nixos/modules/services/networking/miniupnpd.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        type = types.bool;
+        default = false;
+        description = "Enable the MiniUPnP daemon.";
+      };
+
+      externalInterface = mkOption {
+        type = types.str;
+        description = ''
+          Name of the external interface
```
Comment (jagajaga) [inline nixos/modules/services/networking/miniupnpd.nix]: You forgot dot at the end.

File: nixos/modules/services/networking/miniupnpd.nix [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      internalIPs = mkOption {
+        type = types.listOf types.str;
+        example = [ "192.168.1.0/24" ];
+        description = ''
+          The IP address ranges to listen on.
+        '';
+      };
+
+      natpmp = mkOption {
+        default = false;
```
Comment (jagajaga) [inline nixos/modules/services/networking/miniupnpd.nix]: Please use `mkEnableOption`.

File: nixos/modules/services/networking/miniupnpd.nix [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        description = ''
+          Whether to use to enable NAT-PMP support
+        '';
+      };
+
+      upnp = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          Whether to use to enable UPNP support
```
Comment (jagajaga) [inline nixos/modules/services/networking/miniupnpd.nix]: Another dot forgotten :)

**Comment by jgillich** [comment]
@jagajaga Thanks, based all of this on another service so I had no idea `mkEnableOption` even exists. Is there documentation for them or where can I find them in the source tree?

Anyway, all fixed, also fixed the descriptions that made no sense.

File: nixos/modules/services/networking/miniupnpd.nix [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      listening_ip=${range}
+    '') cfg.internalIPs}
+
+    ${cfg.appendConfig}
+  '';
+in
+{
+  options = {
+    services.miniupnpd = {
+      enable = mkEnableOption "Enable the MiniUPnP daemon.";
```
Comment (jagajaga) [inline nixos/modules/services/networking/miniupnpd.nix]: As an argument you need to write just the name of option :)
In this case you need to write just the name of service.
You can search through nixpkgs and find some examples.

**Comment by jgillich** [comment]
Oh I did look at examples, but seem to have missed that. ;)

**Comment by jgillich** [comment]
Btw why is there no `mkDisableOption` as a shorthand for `default = true`?

**Comment by jgillich** [comment]
Please don't merge yet, the `-d` option is really verbose, I need to change it to use the PID file.

**Comment by zimbatm** [comment]
Is it possible to run miniupnpd in it's own user instead of root ?

**Comment by jgillich** [comment]
@zimbatm You tell me haha. This thing has bascially zero documentation and I'm not exactly an expert in upnp.

**Comment by zimbatm** [comment]
@jgillich Looking at Gentoo again, it seems like it's touching iptables so probably needs root access. Their init script also runs a `/etc/miniupnpd/iptables_removeall.sh` script on stop which might apply for us as well ? See https://gitweb.gentoo.org/repo/gentoo.git/tree/net-misc/miniupnpd/files/miniupnpd-init.d

**Comment by zimbatm** [comment]
ArchLinux seem to be doing the same: https://projects.archlinux.org/svntogit/community.git/tree/trunk/miniupnpd.systemd?h=packages/miniupnpd

**Comment by jgillich** [comment]
They also call [a script to create the rules](https://github.com/miniupnp/miniupnp/blob/master/miniupnpd/netfilter/iptables_init.sh) before starting the service, I'm currently doing that via `firewall.extraCommands`. That could probably be integrated into the service, I'll also try running it as non-root.

**Comment by zimbatm** [comment]
Yeah, I wonder how it interacts with the NixOS built-in firewall, are the ruleset not going to overlap ?

**Comment by jgillich** [comment]
By "them" I meant Arch and Gentoo. :) miniupnpd only uses the chains set via `upnp_forward_chain` and `upnp_nat_chain` in the config and doesn't touch the firewall.

**Comment by zimbatm** [comment]
:+1:

**Pull request merged**